### PR TITLE
Initial implementation of ANRv2

### PR DESCRIPTION
**Pull request by romtsn**
_#skip-changelog_

## :scroll: Description

* Introduces new pipeline for reporting historical events that should be backfilled with scope/options data that has been serialized to disk in the previous application run (Backfillable hint and BackfillingEventProcessor)
* Introduces PersistingScopeObserver/PersistingOptionsObserver that persist scope and option values to disk and get notified whenever any of the observed data changes
* New AnrV2Integration which wroks for API 30 and above and uses [getHistoricalProcessExitReasons](https://developer.android.com/reference/kotlin/android/app/ActivityManager#getHistoricalProcessExitReasons(kotlin.String,%20kotlin.Int,%20kotlin.Int)) to report ANRs
	* Only the latest ANR being reported enriched with the serialized scope/options data from the previous run
	* All other historical ANRs are still reported, but without enriching (only static data, like device params)
* New AnrV2EventProcessor which takes care of backfilling/filling contexts/scope data for the ANRv2
* As the historical exits list stores values in a ring buffer and does not remove them upon reading, we have to keep track of the last reported ANR timestamp - this is done in AndroidEnvelopeCache, where we store this timestamp in a marker file under cacheDir
* Extracted common things to ContextUtils that are used both in AnrV2EventProcessor and DefaultAndroidEventProcessor
* Added a bunch of `equals`/`hashCode` for the protocol classes to be able to assert against them in tests
* Made `UncaughtExceptionHint` public, because we were using `DiskFlushNotification` as a proxy for handling uncaught exceptions logic, now it's more explicit (and we also use `DiskFlushNotification` for ANR events)

## :bulb: Motivation and Context

#1796 

## :green_heart: How did you test it?
Automated and manually

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

## :crystal_ball: Next steps
The PR is targeting integration branch to simplify PR review process. Other PRs to follow:
* Make sure last known TX is correctly associated with the latest ANR
* Mark last known session as abnormal
* Parse the thread dump and transform it into sentry threads payload

### CODE REVIEW

File: sentry-android-core/src/main/java/io/sentry/android/core/AnrV2Integration.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+      if (isBackground) {
+        message = "Background " + message;
+      }
+
+      // TODO: here we should actually parse the trace file and extract the thread dump from there
+      // and then we could properly get the main thread stracktrace and construct a proper exception
+      final ApplicationNotResponding error =
+          new ApplicationNotResponding(message, Looper.getMainLooper().getThread());
+      final Mechanism mechanism = new Mechanism();
+      mechanism.setType("ANRv2");
```
Comment (romtsn) [inline sentry-android-core/src/main/java/io/sentry/android/core/AnrV2Integration.java:244]: Not sure if we should keep it as `ANRv2` or change it to `ANR`? This way we can query on looker by mechanism I think and see who's sending new ANR events. But also, it'd be possible through the integration list.

The question here is rather: do the users care about this and want to see if the ANR has been reported using new mechanism?

**Comment by romtsn** [comment]
@markushi re. serializing collections - I've checked, and actually all of our collections on scope are concurrent already (event `Contexts` class, which itself extends `ConcurrentHashMap`), so I think we're good here.

**Review by markushi — APPROVED** [review]
Nice one, the earlier walkthrough definitely helped a lot when reviewing! I added a few minor comments about `final` keywords and one suggestion for the JSON serializer APIs.

A few more thoughts:
1. I think we need to ensure that the same ANR won't be sent over and over again, ultimately DDoS-ing us in case something goes wrong along the whole process. I can see the marker file gets written in the `AndroidEnvelopeCache` after `super.store(...)` is called. Could it make sense to reverse the order here? Like first save the marker file and then store the envelope? If storing the envelope fails we'd loose the event, but we'd be safe from reporting the same ANR on every app start.

2. The mechanism which ensures an ANR event is enriched with old disk data while no new data is written is A) adding the ANR integration at the beginning and B) having a single threaded executer service right?

File: sentry/src/main/java/io/sentry/JsonSerializer.java [hunk 2]
```diff
@@ -116,14 +117,29 @@ public JsonSerializer(@NotNull SentryOptions options) {
 
   // Deserialize
 
+  @SuppressWarnings("unchecked")
   @Override
-  public <T> @Nullable T deserialize(@NotNull Reader reader, @NotNull Class<T> clazz) {
+  public <T, R> @Nullable T deserialize(
+      @NotNull Reader reader,
+      @NotNull Class<T> clazz,
+      @Nullable JsonDeserializer<R> elementDeserializer) {
```
[the remark this answers is not in the document]
Comment (romtsn) [reply sentry/src/main/java/io/sentry/JsonSerializer.java:125]: yeah, a separate method is actually a better idea, I'll change that 👍

**Comment by romtsn** [comment]
> The mechanism which ensures an ANR event is enriched with old disk data while no new data is written is A) adding the ANR integration at the beginning and B) having a single threaded executer service right?

Yes, that's right. Even though it's a bit brittle, I've added a super sophisticated integration test in https://github.com/getsentry/sentry-java/pull/2549/commits/6c51875b8c0ad1ccde74e1fb7304239ee6d68e7d, but I think this one is necessary as it tests exactly this scenario end-to-end. Not sure if we can do something better for now without changing even more in the pipeline which I'd really not want to do :D

**Pull request merged**